File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Decorator/Decorator.py
from abc import ABC,abstractmethod
class Beverage:
    def __init__(self):
        pass
    @abstractmethod
    def getDescription(self) -> str:
        pass
    @abstractmethod
    def getCost(self)->int:
        pass

class Coffee(Beverage):
    def __init__(self):
        pass
    def getCost(self):
        return 1.5
    def getDescription(self):
        return "Coffee"

class CondimentDecorator(Beverage):
    def __init__(self,beverage):
        super().__init__(beverage)
    def getCost(self):
        return self.beverage.getCost()
    def getDescription(self):
        return self.beverage.getDescription() + ","

class Milk(CondimentDecorator):
    def __init__(self,beverage):
        super().__init__(beverage)
    def getCost(self):
        return self.beverage.getCost() + 0.5
    def getDescription(self):
        return self.beverage.getDescription() + "MILK"

class Sugar(CondimentDecorator):
    def __init__(self,beverage):
        super().__init__(beverage)
    def getCost(self):
        return self.beverage.getCost() +0.25
    def getDescription(self):
        return self.beverage.getDescription() + "Sugar"

beverage = Coffee()
print(beverage.getDescription() + "-$" + beverage.getCost())
beverage = Milk(beverage)
print(beverage.getDescription() + "-$" + beverage.getCost())
beverage = Sugar(beverage)
print(beverage.getDescription() + "-$" + beverage.getCost())

File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Decorator/Decorator.ts
abstract class Beverage {
  abstract getDescription(): string;
  abstract getCost(): number;
}
class Coffee extends Beverage {
  override getCost(): number {
    return 1.5;
  }
  override getDescription(): string {
    return "Coffee";
  }
}
class CondimentDecorator extends Beverage {
  protected beverage: Beverage;
  constructor(beverage: Beverage) {
    super();
    this.beverage = beverage;
  }
  override getCost(): number {
    return this.beverage.getCost();
  }
  override getDescription(): string {
    return this.beverage.getDescription() + ",";
  }
}
class Milk extends CondimentDecorator {
  constructor(beverage: Beverage) {
    super(beverage);
  }
  getCost(): number {
    return this.beverage.getCost() + 0.5;
  }
  getDescription(): string {
    return this.beverage.getDescription() + "MILk";
  }
}
class Sugar extends CondimentDecorator {
  constructor(beverage: Beverage) {
    super(beverage);
  }
  getCost(): number {
    return this.beverage.getCost() + 0.25;
  }
  getDescription(): string {
    return this.beverage.getDescription() + "Sugar";
  }
}
let beverage = new Coffee();
console.log(beverage.getDescription() + "-$" + beverage.getCost());
beverage = new Milk(beverage);
console.log(beverage.getDescription() + "-$" + beverage.getCost());
beverage = new Sugar(beverage);
console.log(beverage.getDescription() + "-$" + beverage.getCost());
